rules/DeadCode/Rector/FunctionLike/TooWideReturnTypeRector.php

```diff
@@ -0,0 +1,243 @@
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Narrow too wide return type declarations if possible',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    public function foo(): string|int|\DateTime
+    {
+        if (rand(0, 1)) {
+            return 'text';
+        }
+
+        return 1000;
+    }
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    public function foo(): string|int
+    {
+        if (rand(0, 1)) {
+            return 'text';
+        }
+
+        return 1000;
+    }
+}
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class, Function_::class, Closure::class, ArrowFunction::class];
+    }
+
+    /**
+     * @param ClassMethod|Function_|Closure|ArrowFunction $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $scope = ScopeFetcher::fetch($node);
+
+        if ($this->shouldSkipNode($node, $scope)) {
+            return null;
+        }
+
+        $returnStatements = $this->betterNodeFinder->findInstanceOf($node, Return_::class);
+        $isAlwaysTerminating = $node instanceof ArrowFunction
+            || $this->terminatedNodeAnalyzer->isAlwaysTerminating($node);
+
+        if ($returnStatements === [] && ! $node instanceof ArrowFunction) {
+            $node->returnType = $isAlwaysTerminating
+                ? new Identifier('never')
+                : new Identifier('void');
+
+            return $node;
+        }
+
+        $returnType = $node->returnType;
+
+        if (! $returnType instanceof UnionType) {
+            return null;
+        }
+
+        $actualReturnTypes = $this->collectActualReturnTypes($node, $returnStatements, $isAlwaysTerminating);
+        $newReturnType = $this->narrowUnionReturnType($returnType, $actualReturnTypes);
+
+        if ($newReturnType === null) {
+            return null;
+        }
+
+        $node->returnType = $newReturnType;
+
+        return $node;
+    }
+
+    private function shouldSkipNode(ClassMethod|Function_|Closure|ArrowFunction $node, Scope $scope): bool
+    {
+        if ($this->hasYield($node)) {
+            return true;
+        }
+
+        if (! $node instanceof ClassMethod) {
+            return false;
+        }
+
+        if ($node->isPrivate() || $node->isFinal()) {
+            return false;
+        }
+
+        if ($node->isAbstract()) {
+            return true;
+        }
+
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+
+        if (! $classReflection instanceof ClassReflection) {
+            return true;
+        }
+
+        return ! $classReflection->isFinal();
```

Comment:
use `isFinalByKeyword()` to avoid check by docblock:

```suggestion
        return ! $classReflection->isFinalByKeyword();
```

Reply:
I don't see why a method marked as `@final` via the keyword can't be updated? The intention of the developer is that the method should be final

Reply:
we use that in other rules, apply only on native `final`

Reply:
This rector seems to use `->isFinal()`:

https://github.com/rectorphp/rector-src/blob/51dc24ca61e227e9f416886c69471bdaca0a7034/rules/Php80/Rector/ClassMethod/FinalPrivateToPrivateVisibilityRector.php#L81-L83

Reply:
Ah, that makes sense